### PR TITLE
Pubdev 3728 glrm binary loss numeric

### PR DESCRIPTION
**Pull request by wendycwong**
1. Added code to support using logistic loss for binary columns read in as either categorical or numeric for GLRM.

2. Added pyunit tests to check that our new addition works for all init modes.  In addition, we compare the glrm singular values calculated when the binary columns are read in as categorical and numerical to make sure I have implemented the change correctly.

3.  Added Runit test from Megan/User to make sure we have fixed their problem.

Truly appreciate any help/comments you may have.

-- Wendy

### CODE REVIEW

**Comment by wendycwong** [comment]
@h2o-ops please test!

**Comment by abal5** [comment]
@h2o-ops please test!

**Comment by wendycwong** [comment]
@h2o-ops please test

**Comment by surekhajadhwani** [comment]
@h2o-ops please test!

**Comment by wendycwong** [comment]
From my personal Jenkins run, I have two Jenkins errors.  All other tests passed.  Check here: http://172.17.2.151:8080/view/wendy/

Not sure what is going on here.....

**Comment by arnocandel** [comment]
@h2o-ops please test!

File: h2o-algos/src/main/java/hex/DataInfo.java [hunk 1]
```diff
@@ -747,6 +747,66 @@ public final int nextNumericIdx(int currentColIdx) {
...
+    final Vec [] vecs = _adaptedFrame.vecs();
+
+    // first do all of the expanded categorical names
+    for(int i = 0; i < _cats; ++i) {
+      for (int j = (_useAllFactorLevels || vecs[i] instanceof InteractionWrappedVec) ? 0 : 1; j < vecs[i].domain().length; ++j) {
+        int jj = getCategoricalId(i, j);
+        if(jj < 0)
+          continue;
+
+        if ((_catOffsets[i+1]-_catOffsets[i]) == 1) {
```
Comment (tomasnykodym) [inline h2o-algos/src/main/java/hex/DataInfo.java]: This should be moved outside of the inner loop. It's really like this.

Comment (wendycwong) [inline h2o-algos/src/main/java/hex/DataInfo.java]: Tomas:

Done!  It should be more efficient now.  Thanks.

File: h2o-py/tests/testdir_algos/glrm/pyunit_glrm_PUBDEV_3728_binary_numeric.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+  # exercise logistic loss with numeric columns
+  glrm_h2o_num = H2OGeneralizedLowRankEstimator(k=5, loss_by_col=loss_all, recover_svd=True, transform="STANDARDIZE",
+                                              seed=12345, init="User", user_y=initial_y_h2o)
+  glrm_h2o_num.train(x=prostateF_num.names, training_frame=prostateF_num, validation_frame=prostateF_num)
+  glrm_h2o_num.show()
+
+  # singular values from glrm models should equal if binary columns with binary loss are read in as either
+  # categorical or numerics.  If not, something is wrong.
+  assert pyunit_utils.equal_two_arrays(glrm_h2o._model_json["output"]["singular_vals"],
```
Comment (meganjkurka) [inline h2o-py/tests/testdir_algos/glrm/pyunit_glrm_PUBDEV_3728_binary_numeric.py]: It looks like this is comparing the same model to itself (both are `glrm_h2o`).  I think one should be changed to `glrm_h2o_num`.

Comment (wendycwong) [inline h2o-py/tests/testdir_algos/glrm/pyunit_glrm_PUBDEV_3728_binary_numeric.py]: Hi Megan:

Good catch.  Fixed it.  Thanks.  Wendy

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 3]
```diff
@@ -241,16 +241,26 @@ private void initLoss() {
...
       Vec vi = _train.vec(i);
       GlrmLoss lossi = _lossFunc[i];
-      if (vi.isNumeric()) {
+        // take care of pure numeric columns.  Binary numeric columns has isBinary() = true
+      if (vi.isNumeric() && !vi.isBinary()) {
         if (!lossi.isForNumeric())
           error("_loss_by_col", "Loss function " + lossi + " cannot be applied to numeric column " + i);
+      }  else if (vi.isBinary()) {
+        // Allow numeric loss functions, multi-loss on binary columns too; but not the other way round!
+        if (!lossi.isForBinary() && !lossi.isForNumeric() && !lossi.isForCategorical())
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: This should be just `!lossi.isForBinary()`, per description of the method. You may need to tweak the loss functions enum so that `isForBinary()` returns true for all losses that are suitable for binary columns

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 4]
```diff
@@ -241,16 +241,26 @@ private void initLoss() {
...
+          error("_loss_by_col", "Loss function " + lossi + " cannot be applied to binary column " + i);
+        // make sure we are not calling init mode SVD for categorical binary columns with binary loss
+        if (vi.isCategorical() && lossi.isForBinary()) {
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: I don't understand this check. Isn't categorical binary exactly the same as numeric binary (with the exception of the `type` byte)?

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 5]
```diff
@@ -652,22 +662,31 @@ public void computeImpl() {
...
         tempinfo = new DataInfo(_train, null, 0, true, _parms._transform, DataInfo.TransformType.NONE,
                 false, false, false, /* weights */ false, /* offset */ false, /* fold */ false);
  //       DKV.put(tempinfo._key, tempinfo);
 
+        model._output._permutation = tinfo._permutation;
+        model._output._nnums = tinfo._nums;
+        model._output._ncats = tinfo._cats;
+        model._output._catOffsets = tinfo._catOffsets;
+        int[] numLevels = tinfo._adaptedFrame.cardinality();
+        int[] numCatColumns = tinfo._adaptedFrame.cardinality();
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: why do you need to create the same array twice?

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java [hunk 6]
```diff
@@ -106,25 +106,49 @@
 
 
   //--------------------------------------------------------------------------------------------------------------------
-  // Loss functions for binary features
+  // Loss functions for binary features read in as categorical or numeric
   //--------------------------------------------------------------------------------------------------------------------
 
   Logistic {
-    @Override public boolean isForNumeric() { return false; }
+    @Override public boolean isForNumeric() { return true; }
```
Comment (st-pasha) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java]: Logistic loss cannot be applied to numeric columns! It's only defined for a = 0/1

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java [hunk 7]
```diff
@@ -143,6 +167,10 @@
     @Override public double impute(double u) {
       return u > 0? 1 : 0;
     }
+
+    @Override public int mimpute(double[] u) {
+      return ArrayUtils.maxIndex(u);
+    }
```
Comment (st-pasha) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java]: `mimpute` should be defined in conjunction with `mloss` and `mlgrad`, for categorical loss functions only

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java [hunk 8]
```diff
@@ -207,10 +235,12 @@
     @Override public double[] mlgrad(double[] u, int a) {
       if (!(a >= 0 && a < u.length)) throw new IndexOutOfBoundsException("a must be between 0 and " + (u.length - 1));
```
Comment (st-pasha) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java]: don't need this check then

File: h2o-py/tests/testdir_algos/glrm/pyunit_glrm_PUBDEV_3728_binary_numeric.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+# This unit test is to make sure that logistic loss works with both binary columns read in as either
+# enum or numerics.  In addition, we check that logistic loss work with different init modes for GLRM,
+# namely plusplus (default), user, SVD, random.
+#
+# When the init mode is set to user, we have total control of the initial values of X and Y into the GLRM.  In
+# this case, the singular values obtained for GLRM when binary columns are read as numeric or enum should be the same.
+# This is what I use to make sure my logistic implementation is correct.  We will compare the singular values
+# and throw an error when this is not the case.  For other init modes, the values are close but not equal.
+def glrm_pubdev_3728_arrest():
+  print("Importing prostate.csv data...")
```
Comment (st-pasha) [inline h2o-py/tests/testdir_algos/glrm/pyunit_glrm_PUBDEV_3728_binary_numeric.py]: the default for Python is to use 4-space indentation

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java (hunk 6 above)
Comment (wendycwong) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java]: Agreed.  Logistic should only be for binary columns.

**Comment by wendycwong** [comment]
@h2o-ops please test!

File: h2o-algos/src/main/java/hex/glrm/GLRM.java [hunk 10]
```diff
@@ -241,16 +245,26 @@ private void initLoss() {
...
-      } else if (vi.isBinary()) {
-        // Allow numeric loss functions on binary columns too; but not the other way round!
-        if (!lossi.isForBinary() && !lossi.isForNumeric())
-          error("_loss_by_col", "Loss function " + lossi + " cannot be applied to binary column " + i);
+
+      if (vi.isNumeric())   { // numeric columns
+        if (!vi.isBinary()) { // non-binary numeric columns
+          if (!lossi.isForNumeric())
+            error("_loss_by_col", "Loss function "+lossi+" cannot be applied to numeric column "+i);
+        } else {  // binary numeric columns
```
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: why no `_binaryColumnIndices.add()` here?

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java [hunk 11]
```diff
@@ -120,7 +120,8 @@
     }
     @Override public double lgrad(double u, double a) {
       double s = 1 - 2*a;
-      return s/(1 + Math.exp(s*u));
+      double expsu = Math.exp(s*u);
+      return s*expsu/(1 + expsu);
```
Comment (st-pasha) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java]: or even simpler: `s/(1 + Math.exp(-s*u))`

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java [hunk 12]
```diff
@@ -190,13 +191,13 @@
...
-    @Override public double mloss(double[] u, int a) {
-      if (!(a >= 0 && a < u.length)) throw new IndexOutOfBoundsException("a must be between 0 and " + (u.length - 1));
+    @Override public double mloss(double[] u, int a) { return mloss(u, a, u.length); }
+/*      if (!(a >= 0 && a < u.length)) throw new IndexOutOfBoundsException("a must be between 0 and " + (u.length - 1));
       double sum = 0;
       for (int i = 0; i < u.length - 1; i++)
         sum += a > i ? Math.max(1 - u[i], 0) : 1;
       return sum;
-    }
+    } */
```
Comment (st-pasha) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java]: you can probably delete this commented out code (and in line 212 too)

File: h2o-algos/src/main/java/hex/glrm/GLRM.java (hunk 10 above)
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: In this case the column vi is already numerical.  There is no need to change to numerical later.

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java (hunk 11 above)
Comment (wendycwong) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java]: Super sleek!  Thanks.

File: h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java (hunk 12 above)
Comment (wendycwong) [inline h2o-genmodel/src/main/java/hex/genmodel/algos/glrm/GlrmLoss.java]: Done!

File: h2o-algos/src/main/java/hex/glrm/GLRM.java (hunk 10 above)
Comment (st-pasha) [inline h2o-algos/src/main/java/hex/glrm/GLRM.java]: hmm, the comment on line 64 describes it as "binary columns using binary loss functions." Perhaps the description could be clarified then?

Also, I'd switch the order of then and else clauses here -- right now it looks like a double negative...
